Univa/rumcake
# Pointing device

## Applicable Changes



- [ ] Modification to behaviour of existing functionality (potential breaking change)
- [ ] Modification to existing public APIs (potential breaking change)
- [ ] User docs modification
- [ ] Rust toolchain updates
- [ ] Workflow change (GitHub Actions)
- [ ] Repository housekeeping (e.g. READMEs, licensing, templates, etc.)

## Related Issues



#8 

## Description of Changes



Added a `pointing` module to support pointing devices like a mouse or touchpad. Note that this only adds support for pointing devices over USB for now. Bluetooth support to come later.

## Testing Instructions



Try out the IQS5xx pointing driver in your `main.rs` (assuming a non-split keyboard):

```rust
#[keyboard(
    usb,
    pointer(
        driver_setup_fn = test_pointer,
    )
)]
struct MyKeyboard;

impl rumcake::pointer::PointingDevice for MyKeyboard {
    type MouseEventCollector = Self;
}
impl rumcake::pointer::MouseEventCollector for MyKeyboard {}
async fn test_pointer() -> impl rumcake::pointer::PointingDriver {
    rumcake::drivers::iqs5xx::setup_driver(
        setup_i2c! {
            /* ... */
        },
        input_pin!( /* pin identifier */ ),
        output_pin!( /* pin identifier */ ),
    )
}
```

The touchpad should be capable of basic cursor movements, scrolling (with two fingers sliding), left-clicks (via single finger tap), right-clicks (via two finger tap), and left-click holding (press hold).

More complicated behaviour can be implemented by overriding the implementation of `tick` and `handle_event` in the `IQS5xxPointingDriver` trait

## Check-list

- [ ] I have updated or added to the user docs where applicable
- [ ] I have tested the changes on real hardware if applicable
- [x] The changes do not break compilations with other potentially related feature flags
- [x] Credit given where applicable
